### ISSUE TITLE
[sql] Scoping Still Broken on `extend`

### ISSUE THREAD

**Issue by belph** [brownplt/pyret-lang#733]
Consider the `extend` keyword's syntax:
```
extend <TABLE> using <COLS...>:
  <NEW-COL>: <REDUCER-EXPR> of <COL>
end
```
Now, `<COL>` needs to be one of the columns listed in `<COLS...>`, but this is currently not enforced. If `<COL>` is out of scope, there is an error (as expected), but the following code raises an `"Impossible"` error:
```
running-sum = { ... }
x = 12
extend tbl using b:
  c: running-sum of x
end
```

I'm posting a bug report since I am unsure where this check for scoping should go :)

My gut tells me that this should be a well-formedness error, but is that the appropriate place to check for scoping issues?

**Comment by jpolitz** [comment]
This belongs in the environment-tracking visitor in `resolve-scope.arr`, which needs to have a case that describes the binding structure of `extend`.  It's a big `map` over the AST that resolves each identifier by tracking an environment from names to unique `s-atom`s.  Unbound ids are left as `s-global`s, and a later step checks that each global variable is provided by some imported module, and reports an unbound-id if not.

**Comment by belph** [comment]
@jpolitz The only issue I see with that is the provided code snippet will then say something like "`x` is unbound," to which users will say, "No it isn't!"

**Comment by jpolitz** [comment]
Thanks for explaining, I misunderstood and was thinking of the reducer case like other extend clauses, as an expression position.

This ought to be a well-formedness error.

**Comment by blerner** [comment]
What is "this" that ought to be the well-formedness error?  Mentioning an external identifier?  Definitely not. Reducing over a name that isn't declared as a column?  Maybe...

**Comment by jpolitz** [comment]
The grammar for reducers takes a `NAME` after `of`, not an expression, and the `NAME` is expected to be a column name.

If it were an expression, the two cases you mention would overlap, no?

**Comment by blerner** [comment]
Yes, then I agree.  I just wanted to make sure we didn't somehow over-constrain the innards of table expressions to be so tightly scoped as to only work over their own data!  But saying `reducer := <expr> OF NAME`, where the name must be one of the column names declared, sounds fine to me.

**Comment by belph** [comment]
Fixed by  dceb15a.
